django/core/handlers/wsgi.py

```diff
@@ -12,56 +12,46 @@
 _slashes_re = _lazy_re_compile(rb"/+")
 
 
-class LimitedStream:
-    """Wrap another stream to disallow reading it past a number of bytes."""
+class LimitedStream(IOBase):
+    """
+    Wrap another stream to disallow reading it past a number of bytes.
+
+    Based on the implementation from werkzeug.wsgi.LimitedStream
+    See https://github.com/pallets/werkzeug/blob/dbf78f67/src/werkzeug/wsgi.py#L828
+    """
 
     def __init__(self, stream, limit):
-        self.stream = stream
-        self.remaining = limit
-        self.buffer = b""
-
-    def _read_limited(self, size=None):
-        if size is None or size > self.remaining:
-            size = self.remaining
-        if size == 0:
+        self._read = stream.read
+        self._readline = stream.readline
+        self._pos = 0
+        self.limit = limit
+
+    def read(self, size=-1, /):
+        _pos = self._pos
+        limit = self.limit
+        if _pos >= limit:
             return b""
-        result = self.stream.read(size)
-        self.remaining -= len(result)
-        return result
-
-    def read(self, size=None):
-        if size is None:
-            result = self.buffer + self._read_limited()
-            self.buffer = b""
-        elif size < len(self.buffer):
-            result = self.buffer[:size]
-            self.buffer = self.buffer[size:]
-        else:  # size >= len(self.buffer)
-            result = self.buffer + self._read_limited(size - len(self.buffer))
-            self.buffer = b""
-        return result
-
-    def readline(self, size=None):
-        while b"\n" not in self.buffer and (size is None or len(self.buffer) < size):
-            if size:
-                # since size is not None here, len(self.buffer) < size
-                chunk = self._read_limited(size - len(self.buffer))
-            else:
-                chunk = self._read_limited()
-            if not chunk:
-                break
-            self.buffer += chunk
-        sio = BytesIO(self.buffer)
-        if size:
-            line = sio.readline(size)
+        if size == -1 or size is None:
+            size = limit - _pos
         else:
-            line = sio.readline()
-        self.buffer = sio.read()
+            size = min(size, limit - _pos)
+        data = self._read(size)
+        self._pos += len(data)
+        return data
+
+    def readline(self, size=-1, /):
+        _pos = self._pos
+        limit = self.limit
+        if _pos >= limit:
+            return b""
+        if size == -1 or size is None:
+            size = limit - _pos
+        else:
+            size = min(size, limit - _pos)
+        line = self._readline(size)
+        self._pos += len(line)
         return line
 
-    def close(self):
```

Comment:
What is the consequence of removing `close`? What does `close` on `IOBase` actually do?

Reply:
I'll double check. Not much, but more than `pass` IIRC. Likely just checking whether it's already closed.

Reply:
[`IOBase.close()`](https://github.com/python/cpython/blob/ad40764f5b1e0a11a8b4a30b7ce2dacc20fda12f/Lib/_pyio.py#L412-L421) checks whether a flag is set and, if not, calls `self.flush()` which, in the case of [`IOBase.flush()`](https://github.com/python/cpython/blob/ad40764f5b1e0a11a8b4a30b7ce2dacc20fda12f/Lib/_pyio.py#L402-L408), calls [`self._checkClosed()`](https://github.com/python/cpython/blob/ad40764f5b1e0a11a8b4a30b7ce2dacc20fda12f/Lib/_pyio.py#L501-L506). That checks the `self.closed` property which just returns the same `self.__closed` flag and raises a `ValueError` if we're already closed.

So the implementation basically doesn't do anything when `close()` is called for the first time, but raises an exception if we try to close something already closed. Given that the stream shouldn't really be closed twice - that'd be a bug somewhere - I think this is fine. (Besides, `werkzeug` does it like this and they'd have had problems otherwise.)

Reply:
Sounds good :)